app/controllers/queue_items_controller.rb

```diff
@@ -0,0 +1,61 @@
+class QueueItemsController < ApplicationController
+  before_filter :require_user
+  
+  def index
+    @queue_items = current_user.queue_items
+  end
+
+  def create
+    video = Video.find(params[:video_id])
+    queue_video(video)
+    redirect_to my_queue_path
+  end
+
+  def destroy
+    queue_item = QueueItem.find(params[:id])
+    queue_item.destroy if current_user.queue_items.include?(queue_item)
+    current_user.normalize_queue_item_positions
+    redirect_to my_queue_path
+  end
+
+  def update_queue
+    begin
+      update_queue_items
+      current_user.normalize_queue_item_positions
+    rescue ActiveRecord::RecordInvalid
+      flash[:danger] = "Invalid position numbers."
+    end
+    redirect_to my_queue_path
+  end
+
+  private 
+
+  def queue_video(video)
+    QueueItem.create(video: video, user: current_user, position: new_queue_item_position) unless current_user_queued_video?(video)
+  end
+
+  def new_queue_item_position 
+    (current_user.queue_items.count + 1)
+  end
+
+  def current_user_queued_video?(video)
```

Comment:
This method should be refactored to User model rather. Think about implementation.